java-checks/src/main/java/org/sonar/java/checks/FunctionalInterfacesSpecialisedCheck.java

```diff
@@ -0,0 +1,265 @@
+      if (reportString == null) {
+        return;
+      }
+      report(newClassInterface, reportString);
+    }
+  }
+
+  @CheckForNull
+  private static String matchInterface(TypeTree interfaceType) {
+    TypeArguments args = interfaceParameters(interfaceType);
+    if (args == null) {
+      return null;
+    }
+
+    String reportString = null;
+    switch (interfaceType.symbolType().fullyQualifiedName()) {
+      case "java.util.function.Function":
+        return handleFunctionInterface(args);
+      case "java.util.function.BiFunction":
+        return handleBiFunctionInterface(args);
+      case "java.util.function.Supplier":
+        reportString = handleOneParameterInterface(args);
+        return reportString != null ? (reportString + "Supplier") : null;
+      case "java.util.function.Consumer":
+        reportString = handleOneParameterInterface(args);
+        return reportString != null ? (reportString + "Consumer") : null;
+      case "java.util.function.BiConsumer":
+        return handleBiConsumerInterface(args);
+      case "java.util.function.Predicate":
+        reportString = handleOneParameterInterface(args);
+        return reportString != null ? (reportString + "Predicate") : null;
+      case "java.util.function.UnaryOperator":
+        reportString = handleOneParameterInterface(args);
+        return reportString != null ? (reportString + "UnaryOperator") : null;
+      case "java.util.function.BinaryOperator":
+        reportString = handleOneParameterInterface(args);
+        return reportString != null ? (reportString + "BinaryOperator") : null;
+      case "java.util.Optional":
+        reportString = handleOneParameterInterface(args);
+        return reportString != null ? ("Optional" + reportString) : null;
+      default:
+        return null;
+    }
+  }
+
+  private static String handleOneParameterInterface(TypeArguments args) {
+    return new ParameterTypeNameAndTreeType(args.get(0)).paramTypeName;
+  }
+
+  private static String handleFunctionInterface(TypeArguments args) {
+    ParameterTypeNameAndTreeType firstArgument = new ParameterTypeNameAndTreeType(args.get(0));
+    ParameterTypeNameAndTreeType secondArgument = new ParameterTypeNameAndTreeType(args.get(1));
+    if (firstArgument.paramTypeName != null && secondArgument.paramTypeName != null) {
+      return firstArgument.paramTypeName + "To" + secondArgument.paramTypeName + "Function";
+    } else if (firstArgument.paramTypeName == null && secondArgument.paramTypeName != null) {
+      return "To" + secondArgument.paramTypeName + "Function<" + firstArgument.originalName + ">";
+    } else if (firstArgument.paramTypeName != null) {
+      return firstArgument.paramTypeName + "Function<" + secondArgument.originalName + ">";
+    } else {
+      return !args.get(0).is(Tree.Kind.PARAMETERIZED_TYPE) ? ("UnaryOperator<" + firstArgument.originalName + ">") : "UnaryOperator<" + firstArgument.originalName;
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Add parentheses to make the operator precedence explicit. ([squid:S864](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS864))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.java%3Ajava&pullRequest=1978&issues=AWKZ5Bfg-RUO4oGVXasB&open=AWKZ5Bfg-RUO4oGVXasB)

[ISSUE_KEY]: AWKZ5Bfg-RUO4oGVXasB